### ISSUE TITLE
Visual Studio 2015 no longer uses a binary reference to a library when its project is loaded in the same solution (works in Visual Studio 2013)

### ISSUE THREAD

**Issue by TyreeJackson** [dotnet/roslyn#4358]
I have a solution with several projects.  The least dependent project, lets call it project A, has some custom post compilation build steps that alter the assembly library before it is used by the other projects in the solution.  Due to this reason, the dll of project A is directly referenced by the other projects that depend on this project, lets call them projects B and C.  Project B simply depends on project A.  Project C is a test project that contains unit tests for project A.

In Visual Studio 2013, projects B and C compile fine, the unit tests run fine and the code coverage metrics report displays the metrics for project A correctly.

However, in Visual Studio 2015, projects B and C appear to be behaving as if they had a project reference to project A instead of the binary reference while project A is included in the same solution.  As a result, several types that are woven into the assembly library during a post compilation build step involving ILAsm.exe, are seen as missing in projects B and C.  This causes projects B and C to fail to compile.  Also note that in Visual Studio 2013 both the project version and the binary version of Project A appear in the Object Browser.  In Visual Studio 2015, only the project version of Project A appears in the Object Browser.  After I remove Project A from the solution, then the project version disappears from the Object Browser as expected and the binary version appears instead.  I believe that **this is a bug** or at least a breaking change worth documenting and hopefully providing a work around.  

If I remove project A from the solution, then projects B and C revert to behaving correctly using the binary reference and compile fine.

But then, here is the catch, the code coverage metrics fail to report at all on project A since it has been removed from the solution even though the tests are passing.  They show up under "External" when the unit tests are grouped by project in the Test Explorer but project A (or external) is missing from the Code Coverage Results.  This is also the same behavior in Visual Studio 2013, so I don't think that his is a bug or any new behavior.

Is there an option, reference property, registry setting, etc, to force Visual Studio 2015 to revert to the Visual Studio 2013 approach to resolving references as indicated in the c# project instead of switching to using project references?

**Comment by jasonmalinowski** [comment]
This was indeed a design change. There isn't an option or registry setting in the released version of Visual Studio 2015 since we didn't anticipate this causing a problem for anyone. We introduced the behavior both to enable refactoring between VB and C# (which internally get converted to file references), and also because we noticed that many projects in the wild were using file references without realizing it and creating a worse experience.

The conversion from file reference back to project reference is done purely by output path matching ([code here](https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs#L880) if you're curious). One immediate workaround that comes to mind would be to change your file references to refer not to the output path of the project but to 'some other' copy. So if your build modifies bin\Debug\Foo.dll and that's the stated output path, you could either have your build rewriting step copy the final rewritten DLL to bin\Debug\Rewritten\Foo.dll, and have the file reference use *that*, or update the output path of the project to be bin\Debug\BeforeRewriting\Foo.dll, and leave your file references as-is.

Longer term we might need to something nicer here, but that should hopefully get you unblocked right away.

**Comment by TyreeJackson** [comment]
@jasonmalinowski 
Thank you for you response Jason!  The dll is modified in the obj\debug folder before the CopyFilesToOutputDirectory step is called.  We used the [IL Support (2013 version)](https://visualstudiogallery.msdn.microsoft.com/530dc77a-a9b0-43bf-9ae2-9498b0ec15da) ([original](https://visualstudiogallery.msdn.microsoft.com/44034a7b-143d-4b51-b7bc-99aa656ba137)) extension to create the project.

**Simply changing the output path to anything else** (I used *bin\debug\PreFinal*) and updating the references to the new location seems to resolve the implicit project reference issue that was preventing the build and keeping the code coverage metrics from working.  This worked even though the reference path and the new output path matched.  However, the *Object Browser* still did not show the binary version of the assembly.  There may be a behavior here that *you* were not expecting.  I wonder if the output path matching conversion only works partially when the output path is *$ProjectDirectory\bin\debug*.

There was one more anomaly that I forgot to mention in the original post, which is that the `EditorBrowsableAttribute` also stopped working with the project referencing since it was never supported via project references and only ever worked if the library containing its use was referenced via its dll (exe, etc).  Like the *Object Browser*, this issue was not resolved by simply changing the output path.  So I added a post build event to copy the dlls from the *bin\debug\Prefinal* back to *bin\debug* and re-updated the dll references back to that folder and now **all four issues are resolved**.  The solution builds, the code coverage report is correct, the *Object Browser* displays both the project version and the binary version of *project A*, and the `EditorBrowsableAttribute` is mostly working<sup>1</sup>.

The Test Explorer still sees the unit tests that target the referenced dll as being External, but that is a very minor annoyance.

<sup>1</sup> On a side note, I will be submitting another issue regarding the EditorBrowsable no longer being honored on methods that hide other methods using the `new` modifier.

Thanks again for your help!

**Comment by jasonmalinowski** [comment]
So I would have expected that the Object Browser is always populated from Visual Studio's view of the project which would never reflect the metadata view of the world. @DustinCampbell will know more here.

@Pilchie, @dpoeschl, how did we change EditorBrowsableAttribute in VS2015 here? I recall some changes of whether or not it took effect within a project to make the logic remotely consistent with itself, but I don't remember the details.

**Comment by Pilchie** [comment]
The EditorBrowsable issue known.  There currently isn't a good compiler API that lets us distinguish this. (cc @jaredpar, this is another case where we need the API I was talking about yesterday)

**Comment by jasonmalinowski** [comment]
What API do we need here?

**Comment by Pilchie** [comment]
@dpoeschl can probably explain it better than me...

**Comment by dpoeschl** [comment]
@jasonmalinowski Adding some quotes from the notes in internal bug 824590:

When the set of symbols to be used for completion is filtered to visible and browsable members, that filtering only takes into account overrides relationships between symbols in that set (but not hiding relationships). This omission can be seen found in [```RemoveOverriddenSymbolsWithinSet```](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/ISymbolExtensions.cs,9b19b8ae52a7c797).

These relationships are calculated in the compiler but it is not public. Fixing this will require ```OverriddenOrHiddenMembers``` being made public or adding a reimplementation of it in the IDE.

Discussed this with @AnthonyDGreen, and we agree that making this API accessible is desirable. Passing this bug to the compiler team to track the work of making the API accessible. 

(*This was a while ago, so @AnthonyDGreen reserves the right to change his mind now* :smile:)

**Comment by Pilchie** [comment]
At this point, I think we should close this bug tracking the unification of references, as I think we want it in general, and there is now a reasonable workaround in terms of changing the output path so that the rewrite doesn't happen in place.
